File: 543-diameter-of-binary-tree/543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        ans = 0
        def longest(node):
            if not node:
                return 0
            
            l = longest(node.left)
            r = longest(node.right)
            nonlocal ans
            ans = max(ans, l + r)
            return max(l, r) + 1
        
        longest(root)
        return ans

File: turingplanet/tree/543.diameter-of-binary-tree.cpp
/*
 * @lc app=leetcode id=543 lang=cpp
 *
 * [543] Diameter of Binary Tree
 */

#include "lib/headers.h"

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
  public:
    int ans = 0;
    int dfs(TreeNode *root)
    {
        if (!root)
            return 0;
        int left  = dfs(root->left);
        int right = dfs(root->right);
        cout << "Working on " << root->val << ":\n";
        cout << "Path of left has " << left << ',';
        cout << "Path of right has " << right << '\n';
        int len = left + right;
        ans     = max(ans, len);
        return 1 + max(left, right);
    }

    int diameterOfBinaryTree(TreeNode *root)
    {
        if (!root)
            return 0;
        dfs(root);
        return ans;
    }
};
// @lc code=end
